Mathlib/GroupTheory/SpecificGroups/Coxeter.lean

```diff
@@ -0,0 +1,377 @@
+/-
+Copyright (c) 2023 Newell Jensen. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Newell Jensen
+-/
+import Mathlib.Data.ENat.Basic
+import Mathlib.Data.Matrix.Notation
+import Mathlib.GroupTheory.OrderOfElement
+import Mathlib.GroupTheory.PresentedGroup
+import Mathlib.LinearAlgebra.Matrix.Symmetric
+
+/-!
+# Coxeter Systems
+
+This file defines Coxeter systems.
+
+It introduces a type class `IsCoxeterGroup` for groups that are isomorphic to a group
+presentation corresponding to a Coxeter matrix which is registered in a Coxeter system.
+
+The finite Coxeter groups are classified as the four infinite families:
+
+* Aₙ, Bₙ, Dₙ, I₂ₘ
+
+And the six exceptional systems:
+
+* E₆, E₇, E₈, F₄, H₃, H₄
+
+## Main definitions
+
+* `Matrix.IsCoxeter` : A matrix `IsCoxeter` if it is a symmetric matrix with ones on
+  the diagonal and off-diagonal elements are greater than or equal to two.
+* `Matrix.CoxeterGroup` : The group presentation corresponding to a Coxeter matrix.
+* `CoxeterSystem` : A structure recording the isomorphism between a group `W` and the group
+  presentation corresponding to a Coxeter matrix, i.e. `Matrix.CoxeterGroup B M`.
+* `IsCoxeterGroup` : A group is a Coxeter group if it is registered in a Coxeter System.
+
+## References
+
+* https://en.wikipedia.org/wiki/Coxeter_group
+
+* [N. Bourbaki, *Lie Groups and Lie Algebras, Chapters 4--6*](bourbaki1968) chapter IV
+  pages 13--15
+
+## Tags
+
+coxeter system, coxeter group
+-/
+
+noncomputable section
+
+variable {B : Type*} [DecidableEq B]
+
+variable (M : Matrix B B ℕ∞)
```

Comment:
```suggestion
variable (M : Matrix B B ℕ)
```
In a Coxeter group, we may want to be able to state that the order of $s_i s_j$ is $M_{i, j}$ regardless whether the order is finite or not. Then, as in mathlib `orderOf` of an element of infinite order is `0`, I would suggest using `ℕ` instead of `ℕ∞`, but then the `off_diagonal` condition should be changed saying `M i j ≠ 1` instead, and a comment about this should be made in the docstring of `Matrix.IsCoxeter`.

Reply:
Actually, I think `ℕ+` would be better, no?

Reply:
Also, here  you say `M i j ≠ 1` but below in your other comment you have `off_diagonal : ∀ i j : B, i ≠ j → 2 ≤ M i j := by aesop`, so I am confused what you really are wanting here.

Reply:
I had two independant suggestions : 1) add `by aesop` 2) `M i j ≠ 1`.

Could you give a *detailed* explanation why `ℕ+` would be better?

Reply:
Well if we were going to do a combination of `M i j ≠ 1` and `Nat`, then technically the identity matrix would be a Coxeter matrix, which isn't true.  That is, all diagonal elements would be equal to 1 and the off diagonal elements would not be equal to 1, which we don't want.  If we are going to go with `Nat` then I think we should keep the `off_diagonal : ∀ i j : B, i ≠ j → 2 ≤ M i j` as this doesn't allow the off diagonal elements to be zero.

Reply:
Again, Bourbaki allows elements on infinite order (which in mathlib are considered of order `0`)...